### ISSUE TITLE
stack unsafety in traverse

### ISSUE THREAD

**Issue by johnynek** [typelevel/cats#1473]
It seems that traverse is easy to blow the stack. Here is an example:

```scala
package cats

import cats.data.StateT
import implicits._

object TravTest {
  def test(size: Int) = {
    type E[T] = Either[String, T]
    type S[T] = StateT[E, Set[Int], T]
    val statev = (0 until size).toVector.traverse[S, Int] { i =>
      for {
        si <- StateT.pure[E, Set[Int], Int](i)
        _ <- StateT.set[E, Set[Int]](Set(i))
      } yield si
    }
    statev.run(Set.empty)
  }
}

```
By adding a `def traverseM` to applicative, we could side-step this because we could implement `traverseM` with `foldM` which can use tailRecM. For instance, in vector:

```scala
      override def traverseM[G[_], A, B](fa: Vector[A])(f: A => G[B])(implicit G: Monad[G]): G[Vector[B]] =
        foldM(fa, Vector.empty[B]) { (vec, a) => G.map(f(a))(vec :+ _) }

```
This is stack safe, and allows us to traverse to any size.

I'm trying to think of some other ways around this, but, especially for `StateT` it seems hard to make a sufficiently lazy `map2Eval` for this to work.

Interestingly, it seems that if you have `F[Eval[T]] => Eval[F[T]]` then you can solve the problem. This seems to keep coming up as a useful construct (it is also sufficient to implement a tailRecM).

**Comment by johnynek** [comment]
cc @non

**Comment by diesalbla** [comment]
I have just tested this code snippet using several versions of `cats`, in an Ammonite REPL. I run `TravTest.test(N)` with N being 1_000, 10_000, or 100_000

Using version 2.0.0, we still get a stack overflow. However, it seems that at least since version 2.2.0, this problem has been solved.

**Comment by rossabaker** [comment]
Thanks, @diesalbla. For posterity, do we have any idea how it was fixed?  I don't see a linked ticket.

**Comment by johnynek** [comment]
I fixed it in #3521

Forgot to link here. Sorry.

**Comment by rossabaker** [comment]
:+1: Thanks for retesting, and thanks for digging up how it was fixed.
